onnxruntime/core/providers/cpu/tensor/expand.cc

```diff
@@ -0,0 +1,157 @@
+  ONNX_CPU_OPERATOR_VERSIONED_TYPED_KERNEL(                                        \
+      Expand,                                                                      \
+      8,                                                                           \
+      12,                                                                          \
+      TYPE,                                                                        \
+      KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<TYPE>()), \
+      Expand<TYPE>);                                                               \
+  ONNX_CPU_OPERATOR_TYPED_KERNEL(                                                  \
+      Expand,                                                                      \
+      13,                                                                          \
+      TYPE,                                                                        \
+      KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<TYPE>()), \
+      Expand<TYPE>);
+
+REG_EXPAND_KERNEL(float)
+REG_EXPAND_KERNEL(double)
+REG_EXPAND_KERNEL(int8_t)
+REG_EXPAND_KERNEL(int16_t)
+REG_EXPAND_KERNEL(int32_t)
+REG_EXPAND_KERNEL(int64_t)
+REG_EXPAND_KERNEL(uint8_t)
+REG_EXPAND_KERNEL(uint16_t)
+REG_EXPAND_KERNEL(uint32_t)
+REG_EXPAND_KERNEL(uint64_t)
+REG_EXPAND_KERNEL(bool)
+REG_EXPAND_KERNEL(MLFloat16)
+
+template <typename T>
+Status Expand<T>::Compute(OpKernelContext* context) const {
+  auto input_tensor = context->Input<Tensor>(0);
+  auto input_data = input_tensor->template Data<T>();
+  auto input_shape = input_tensor->Shape().GetDims();
+
+  auto input_dims = input_shape.data();
+  auto input_dims_size = static_cast<int64_t>(input_shape.size());
+
+  auto shape_tensor = context->Input<Tensor>(1);
+  auto shape_dims = shape_tensor->Data<int64_t>();
+  std::vector<int64_t> output_shape{shape_dims, shape_dims + shape_tensor->Shape().Size()};
+
+  if (input_shape.size() > output_shape.size()) {
+    ORT_THROW("Invalid expand shape");
+  }
+
+  auto input_shape_iter = input_shape.rbegin();
+  auto output_shape_iter = output_shape.rbegin();
+  while (input_shape_iter != input_shape.rend() &&
+         output_shape_iter != output_shape.rend()) {
+    if (*input_shape_iter != *output_shape_iter) {
+      if (1 == *output_shape_iter) {
+        *output_shape_iter = *input_shape_iter;
+      } else if (1 != *input_shape_iter) {
+        ORT_THROW("Invalid expand shape");
+      }
+    }
+    input_shape_iter++;
+    output_shape_iter++;
+  }
+
+  TensorShape output_tensor_shape(output_shape);
+  auto output_tensor = context->Output(0, output_tensor_shape);
+  auto output_data = output_tensor->template MutableData<T>();
+  auto output_dims = output_shape.data();
+  auto output_dims_size = static_cast<int64_t>(output_shape.size());
+  auto max_dims_size = std::max(input_dims_size, output_dims_size);
+  std::unique_ptr<int64_t[]> input_dim_group{new int64_t[max_dims_size]};
```

Comment:
Why not std::array<int64_t,max_dims_size>?

Reply:
Sure we could go with std::array<int64_t,max_dims_size>.

Reply:
Then why not vector?

Reply:
Because we are not expecting the functionalities provided with vectors, there is no dynamic expansion of memory, push back and size() requirements, much less things like assign, insert or swap - we know how much space we need and at which offset we are placing and retrieving the value.